_posts/2020-03-02-tomcat-thread-model.md

```diff
@@ -0,0 +1,96 @@
+* maxConnections - maksymalna liczba połączeń, które mogą być przetwarzane. Po osiągnięciu maksymalnej liczby połączeń, połączenia nadal są przyjmowane i kolejkowane do czasu osiągnięcia limitu wynikającego z parametru acceptCount. Domyślnie maxConnections jest równe maxThreads dla connectora blokującego i 10000 dla nieblokującego.
+* maxThreads - maksymalna liczba wątków obsługujących żądania. Parametr ten oznacza ile żądań może być symultanicznie przetwarzanych przez serwer. Domyślna wartość tego parametru to 200. Biorąc pod uwagę problem, którym zajmujemy się w tym artykule jest to najważniejszy parametr.
+
+## Monitorowanie puli wątków
+
+Aktualna liczba wątków przetwarzających żądania jest obok rozmiaru sterty jednym z najważniejszych parametrów, które powinny być objęte monitoringiem. W tym przypadku monitorujemy atrybut `currentThreadsBusy` mbean'a `Catalina:name="http-bio-8080",type=ThreadPool` (nazwa puli może być inna, np.: http-nio-8080 dla connectora nieblokującego, można to sprawdzić listując wszystkie mbeany lub sprawdzając nazwę wątku w logu). 
+
+Do monitoringu można użyć narzędzia [jolokia](https://jolokia.org/), które udostępnia JMX za pomocą protokołu http. W takiej konfiguracji wystarczy regularnie odpytywać o stan puli wątków poprzez wywołanie http GET:
+`http://HOST:PORT/jolokia/read/Catalina:name="http-bio-8080",type=ThreadPool/currentThreadsBusy`
+
+W odpowiedzi dostajemy jsona, który w polu value zawiera aktualny rozmiar puli wątków 
+```json
+{
+  "request": {
+    "mbean": "Catalina:name=\"http-bio-8080\",type=ThreadPool",
+    "attribute": "currentThreadsBusy",
+    "type": "read"
+  },
+  "value": 1,
+  "timestamp": 1564763927,
+  "status": 200
+}
+```
+Warto skonfigurować alerting, który będzie ostrzegał o wysycaniu się puli połączeń i pozwoli szybciej zdiagnozować problem.
+
+## Porównanie connectorów blokujących (bio) i nieblokujących (nio)
+
+### Różnica w działaniu
+
+W Tomcacie 7.x domyślną implementacją connectora jest implementacja blokująca. W Tomcacie 8 i wyższych domyślny jest connector nieblokujący. Podstawowa różnica w działaniu polega na tym, że **w connectorze blokującym wątek przypisywany jest do połączenia, a w connectorze nieblokującym do pojedynczego żądania**. Może to mieć duże znaczenie przy zastosowaniu połączeń keep-alive kiedy to jedno połączenie jest wykorzystywane do przesłania wielu żądań i odpowiedzi. W takim modelu wątek na serwerze jest zajęty na cały czas trwania pojedynczego połączenia mimo, że faktyczne przetwarzanie na serwerze zachodzi tylko w obrębie pojedynczego żądania.
+
+### Test
+
+Dla porównania można wykonać prosty test na dwóch rodzajach connectorów. W teście klient raz na sekundę wysyła żądanie do serwera. Łącznie jeden klient wysyła 30 żądań. Uruchamiamy 20 klientów w ciągu 2 sekund. Jednocześnie monitorujemy rozmiar puli wątków.
+
+Tak wygląda wykres czasów odpowiedzi dla connectora blokującego:
```

Comment:
Jak byłby czas to można dodać jeszcze wykres liniowy obrazujący czasy odpowiedzi w zależności od ilości użytkowników. Aż do sytuacji w której pula wątków zostaje wyczerpana. Bo wraz ze wzrostem ilości użytkowników, przewaga jednego nad drugim może jednak wystąpić.

Reply:
Z czasem krucho, ale celowo chciałem pokazać, że nie ma tu jakiejś magii pod spodem, która powoduje, że nagle mamy 100% szybciej, tylko właśnie skupić się na wykorzystanych zasobach.